### PR TITLE
[TRTLLM-7385][feat] Optimize Qwen2/2.5-VL performance

### PR DESCRIPTION
**Pull request by yechank-nvidia**
## Summary by CodeRabbit

- New Features
  - Added multimodal mRoPE support across models, enabling Qwen2/Qwen2.5-VL with configurable rope fusion.
  - Enabled importing HuggingFace Qwen2-VL weights through a new mapping flow.
  - Added selective device transfer for multimodal inputs via path filtering.

- Performance
  - CUDA Graphs path now supports mRoPE with per-batch parameters and non-blocking transfers.

- Improvements
  - Refined vision encoder integration and positional handling.
  - Quickstart example can optionally print context/generation logits and logprobs via flags.

- Tests
  - Added comprehensive Qwen2.5-VL integration tests and a unit test for selective device transfer.

### CODE REVIEW

**Review by chzblych — APPROVED** [review]

**Comment by PerkzZheng** [comment]
The fmha_v2 changes LGTM. I will let others review the rest of the codes. Thanks!

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 1]
```diff
@@ -412,22 +450,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+        seq_lens = []
+        window_index_id = 0
+        vit_merger_window_size = self.window_size // self.spatial_merge_size // self.patch_size
+
+        for grid_t, grid_h, grid_w in grid_thw:
+            llm_grid_h, llm_grid_w = (
+                grid_h // self.spatial_merge_size,
+                grid_w // self.spatial_merge_size,
+            )
+            index = torch.arange(grid_t * llm_grid_h * llm_grid_w).reshape(
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:685]: Just checking: is there no need to explicitly add the `device` argument because it is inferred from `grid_thw`, from which `grid_t` is gotten?

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 2]
```diff
@@ -412,22 +450,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+        hidden_states = hidden_states.reshape(seq_len, -1)
+
+        rotary_pos_emb = rotary_pos_emb.reshape(
+            seq_len // self.spatial_merge_unit, self.spatial_merge_unit, -1)
+        rotary_pos_emb = rotary_pos_emb[window_index, :, :]
+        rotary_pos_emb = rotary_pos_emb.reshape(seq_len, -1)
+        emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
+        position_embeddings = (emb.cos(), emb.sin())
+
+        for layer_num, blk in enumerate(self.blocks):
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py:763]: It was referenced from original HF's code.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 3]
```diff
@@ -318,36 +370,23 @@ def __call__(
 class Qwen2VisionModelBase(nn.Module):
 
     def __init__(self, model_config: ModelConfig[PretrainedConfig],
-                 model_class: type[PreTrainedModel]):
+                 model_class: Union[PreTrainedModel, torch.nn.Module]):
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py]: Fixed. Thx.

File: tensorrt_llm/_torch/modules/rotary_embedding.py [hunk 4]
```diff
@@ -125,3 +125,74 @@ def apply_rotary_pos_emb(q_or_k: torch.Tensor,
         else:
             embed = torch.stack((o1, o2), dim=-1).flatten(-2)
             return torch.cat((embed, q_or_k_pass), dim=-1)
+
+
+class MRotaryEmbedding(RotaryEmbedding):
```
Comment (brb-nv) [inline tensorrt_llm/_torch/modules/rotary_embedding.py:130]: Is there a good way to test this? It'll be shared by other models going forward. So, felt it'll be nice to have a dedicated test.

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 5]
```diff
@@ -183,7 +182,7 @@ def _forward_nope(
                                         attention_mask,
                                         None,
                                         None,
-                                        mrope_config,
+                                        None,
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_llama.py:185]: Will be nice to mention arg names for clarity.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 6]
```diff
@@ -318,36 +369,25 @@ def __call__(
...
+        config = model_config.pretrained_config.vision_config
+        config.torch_dtype = model_config.pretrained_config.torch_dtype
+        self.model_dtype = config.torch_dtype
+
+        if model_class in [
+                Qwen2VisionTransformerPretrainedModel,
+                Qwen2_5_VisionTransformerPretrainedModel
+        ]:
+            # NOTE: For Qwen2VL, we use flash_attention_2 for attention implementation to avoid OOM issue.
+            config._attn_implementation = 'flash_attention_2'
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:384]: I think I'm a bit confused here. Do we use HF's vision implementation?

If we use TRTLLM impl, then why do we need to update this? Why would it matter?

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 7]
```diff
@@ -412,22 +452,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+            max_position_embeddings,
+            bias=True,
+            pos_embd_params=None,
+            rope_fusion=False,
+            layer_idx=layer_idx,
+            dtype=config.torch_dtype,
+            config=model_config,
+        )
+
+    def forward(
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:486]: Can you please leave a comment explaining the main differences here w.r.t base implementation? Is it something to do with rope fusion? What's the incentive to override?

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 8]
```diff
@@ -412,22 +452,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+    # Referenced from transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py
+    @torch.inference_mode()
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        rotary_pos_emb: Optional[torch.Tensor] = None,
+        position_embeddings: Optional[tuple[torch.Tensor, torch.Tensor]] = None,
+        **kwargs,
+    ) -> torch.Tensor:
+        hidden_states = hidden_states + self.attn(
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:543]: Super Nit: Seeing 3 `hidden_states` in the same line is a bit difficult to read. Pls consider separating them out. Not mandatory though.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 9]
```diff
@@ -412,22 +452,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+            self.norm1(hidden_states),
+            rotary_pos_emb=rotary_pos_emb,
+            position_embeddings=position_embeddings,
+            **kwargs,
+        )
+        hidden_states = hidden_states + self.mlp(self.norm2(hidden_states))
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:549]: Super Nit: Same as above.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py [hunk 10]
```diff
@@ -412,22 +452,334 @@ def forward(self, multimodal_params: List[MultimodalParams]):
...
+                num_windows_w,
+                vit_merger_window_size,
+            )
+            index_padded = index_padded.permute(0, 1, 3, 2, 4).reshape(
+                grid_t,
+                num_windows_h * num_windows_w,
+                vit_merger_window_size,
+                vit_merger_window_size,
+            )
+            seqlens = (index_padded != -100).sum([2, 3]).reshape(-1)
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_qwen2vl.py:707]: Nit: Can we read -100 from config? Worst case, may be make it a variable which explains what the value means.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 11]
```diff
@@ -227,12 +240,20 @@ def replay(self, batch_size: int,
         static_tensors["input_ids"][:seqlen].copy_(input_ids)
 
         position_ids = current_inputs["position_ids"]
-        static_tensors["position_ids"][:, :seqlen].copy_(position_ids)
-
-        if "mrope_position_deltas" in current_inputs:
-            assert "mrope_position_deltas" in static_tensors
-            static_tensors["mrope_position_deltas"][:batch_size].copy_(
-                current_inputs["mrope_position_deltas"])
+        print(f"[TRT-LLM DEBUG] position_ids.shape: {position_ids.shape}")
```
Comment (brb-nv) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py:243]: Nit: Let's use logger instead?

**Review by brb-nv — APPROVED** [review]
Mostly nits or questions for my own learning.

**Review by jaedeok-nvidia — APPROVED** [review]
Thanks @yechank-nvidia for optimizing the Qwen VL model. In addition to that, this PR makes the implementation much clearer. I've left several minor comments. Would you please take a look those?

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py (hunk 6 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py:384]: For this single file, we support two architectures, Qwen2-VL & Qwen2.5-VL. I replaced the TRT-LLM impl for Qwen2.5-VL and leave it as HF for Qwen2-VL as it is outdated model.

File: tensorrt_llm/_torch/modules/rotary_embedding.py [hunk 12]
```diff
@@ -125,3 +125,74 @@ def apply_rotary_pos_emb(q_or_k: torch.Tensor,
...
+                for i, m in enumerate(cos.split(self.mrope_section, dim=-1))
+            ],
+                            dim=-1)
+            sin = torch.cat([
+                m[i]
+                for i, m in enumerate(sin.split(self.mrope_section, dim=-1))
+            ],
+                            dim=-1)
+        else:
+            # Fallback to the original RoPE where position_ids is 2D for dummy requests
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/modules/rotary_embedding.py:158]: Only dummy request for create_py_executor(not warmup & Cuda_graph dummy). We expect the 3D shape position_ids for text-only request, too.

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 5 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llama.py]: Done. Thx.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py (hunk 8 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py]: Just did. Thx.

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py (hunk 11 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: This was leftover from my debug print. Removed, thx.

[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: Thx. Removed.

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 13]
```diff
@@ -520,6 +517,20 @@ def add_dummy_requests(
                     for _ in range(max_num_draft_tokens):
                         self.impl.add_token(req_id)
 
+            # NOTE: Planning to get dummy_data from each model. Before that, we need to add dummy mrop_config to the request here.
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: Changed to TODO, thx.

File: tensorrt_llm/inputs/multimodal.py [hunk 14]
```diff
@@ -361,13 +361,18 @@ def to_tensor(self, element: str) -> None:
     def to_device(self,
                   element: str,
                   device: str,
-                  pin_memory: bool = False) -> None:
+                  pin_memory: bool = False,
+                  keyword: Optional[List[str]] = None) -> None:
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/inputs/multimodal.py]: As this was meaning for to include keywords, I changed it as filter_keywords. Or do you have better name for it?

File: tensorrt_llm/inputs/multimodal.py [hunk 15]
```diff
@@ -382,30 +387,90 @@ def to_device(self,
...
+                data, keyword, device, pin_memory)
+        else:
+            # Move all data as before
+            transformed_data = self._apply_tensor_operation(
+                data, "to_device", device=device, pin_memory=pin_memory)
+
         setattr(self, element, transformed_data)
 
+    def _move_multiple_paths_to_device(self, data: Dict[str, Any],
+                                       keyword_list: List[str], device: str,
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/inputs/multimodal.py]: Answered above.

File: tensorrt_llm/_torch/modules/rotary_embedding.py [hunk 16]
```diff
@@ -125,3 +125,74 @@ def apply_rotary_pos_emb(q_or_k: torch.Tensor,
...
+        rope_params: RopeParams,
+        *,
+        head_dim: int,
+        mrope_section: List[int],
+        is_neox: bool = True,
+    ):
+        super().__init__(rope_params, head_dim=head_dim, is_neox=is_neox)
+        self.mrope_section = mrope_section
+
+    def get_cos_sin(self, position_ids: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/modules/rotary_embedding.py]: Thx. Done.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py]: Changed to 'block'. Thx.

File: tensorrt_llm/inputs/multimodal.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jaedeok-nvidia) [reply tensorrt_llm/inputs/multimodal.py]: Oh, I misunderstood. `filter_keywords` looks ok, but I think `target_keywords` (or maybe an even better name 😂) is clearer. The word *filter* might make it unclear if we're including or excluding keywords.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 17]
```diff
@@ -1305,8 +1307,9 @@ def _prepare_tp_inputs(
         num_cached_tokens_per_seq = []  # per sequence
         draft_tokens = []
         draft_lens = []
-        multimodal_params_list = []
         gen_request_seq_slots = []  # per generation request
+        multimodal_params_list = []
+        mrope_position_ids = []
```
[the remark this answers is not in the document]
Comment (jaedeok-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1312]: So, for rope fusion it is needed to keep both variables. Thanks for the explanation.

**Review by mikeiovine — APPROVED** [review]
Runtime changes look good to me

File: tensorrt_llm/inputs/multimodal.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/inputs/multimodal.py]: Changed. Thx.

File: tensorrt_llm/_torch/models/modeling_qwen2vl.py (hunk 10 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_qwen2vl.py]: Sure. Left a comment and note at the very top of the file.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 18]
```diff
@@ -1656,11 +1683,26 @@ def previous_seq_slots_device():
...
-        position_ids = torch.tensor(position_ids,
-                                    dtype=torch.int,
-                                    pin_memory=True)
-        self.position_ids_cuda[:total_num_tokens].copy_(position_ids,
-                                                        non_blocking=True)
+        if self.use_mrope and mrope_position_ids:
+            # NOTE: self.use_mrope is enough for differentiating whether to use mrope_position_ids but
+            # `_create_dummy_context_requests` from `kv_cache_creater` makes an exception that I can not add multimodal_data to the dummy_request
+            # so that we only replace position_ids with mrope_position_ids when it is not a dummy request and for models who is using mrope.
+            mrope_position_ids = torch.cat(mrope_position_ids,
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1690]: Hi @Nekofish-L, thx for raising this issue! Yes, as you pointed out, we found that this cause the memory location discrepancy. We are trying to address this in this [PR](https://github.com/NVIDIA/TensorRT-LLM/pull/7539).

The solution is that we removed `mrope_position_ids` from `multimodal_data_device_paths` which makes it to reside on CPU region.

Thx again for the catch.
